### ISSUE TITLE
ModelCheckpoint save_freq incompatible with Model.fit validation_freq

### ISSUE THREAD

**Issue by bjarthur** [tensorflow/tensorflow#41632]
please see [this issue](https://github.com/keras-team/keras/issues/13689) filed on the now unsupported keras repo.  @gattia there has it [figured out](https://github.com/keras-team/keras/issues/13689#issuecomment-583758348) i believe.  would be nice to have a work around.

in brief, when using `ModelCheckpoint(save_freq=<int>...` combined with `model.fit(validation_freq=<int>...` then "WARNING:tensorflow:Can save best model only with val_loss available, skipping." is emitted. even when you've carefully calculated the two to be the same number of epochs.

**Comment by Saduf2019** [comment]
@bjarthur 
Can you please refer to these links with reported to the error reported;
[link](https://stackoverflow.com/questions/59706714/warningtensorflowearly-stopping-conditioned-on-metric-val-binary-accuracy-wh) [link1](https://stackoverflow.com/questions/53698035/failed-to-get-convolution-algorithm-this-is-probably-because-cudnn-failed-to-in)

Please share simple stand alone code [indented with all dependencies] such that we can replicate the issue faced or if possible share a colab gist with the error for us to analyse.

**Comment by gattia** [comment]
@bjarthur can you please reopen the issue? 

@Saduf2019 the problem is not at the links you've presented. Yes, it originally seems like it could be associated with early stopping, but is not, it is a model checkpoint problem. 

Attached below is a MWE of the problem. I have tested this on TF 2.0.0. locally and 2.3.0 on colab. 

Essentially, if you try to set `save_freq` in `ModelCheckpoint` to be any integer it fails producing the warning: 

`WARNING:tensorflow:Can save best model only with val_loss available, skipping.`

each time that it tries to save the model. 

The logic and explicit code locations that cause this are outlined in this post:
https://github.com/keras-team/keras/issues/13689#issuecomment-662598329

The gist of that post/the problem is that most people using `ModelCheckpoint` are benchmarking it against a validation metric (e.g., `val_loss`), and `val_loss` or any other validation metric is not calculated until after the epoch is completed (it's calculated when `.on_epoch_end`  is evoked). However, by nature of the code logic, `ModelCheckpoint` searches for the `val_loss` at the end of each batch, which is always before the `.on_epoch_end` code is invoked. Therefore, when `ModelCheckpoint` is looking for a `val_loss` it doesn't exist (and can't/won't) so this error will always appear if someone sets save_freq to be anything other than `epoch` which is the default and they want the value being monitored to be a validation metric. 

The example sets up the save_freq to be at the end of an epoch. This will produce the error once per epoch. Though, it can be any value to produce the warning. 

```
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers
from tensorflow.keras.callbacks import ModelCheckpoint
import numpy as np
import tempfile
import os
os.environ['KMP_DUPLICATE_LIB_OK']='True'

# Make Fake Data
x_train = np.random.rand(256, 784).astype("float32")
y_train = np.random.randint(low=0, high=10, size=256)

x_test = np.random.rand(64, 784).astype("float32")
y_test = np.random.randint(low=0, high=10, size=64)

# Build toy model 
inputs = keras.Input(shape=(x_train.shape[1]))
dense = layers.Dense(64, activation="relu")(inputs)
dense = layers.Dense(64, activation="relu")(dense)
outputs = layers.Dense(10)(dense)

model = keras.Model(inputs=inputs, outputs=outputs)

print(model.summary())

model.compile(
    loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
    optimizer=keras.optimizers.Adam(),
    metrics=["accuracy"],
)

# Set Parameters
batch_size = 32 # divisible by number of examples in train (256)
location_save = tempfile.gettempdir() # Get location of temp directory to not store data permanently

# Align modelcheckpoint to end of epoch based on TF version. 
# Note only tf == 2.0 has been tried, as it was what is available on Conda (recommended install method)
if ((tf.__version__[:3] == '2.0') or
	(tf.__version__[:3] == '2.1')
	):
	save_freq = x_train.shape[0]
elif ((tf.__version__[:3] == '2.2') or
	  (tf.__version__[:3] == '2.3')
	  ):
	save_freq = x_train.shape[0] // batch_size

verbose = 2 # To easily see number of epochs and "warning" print statements from TF. 

# Create model checkpoint
model_checkpoint = ModelCheckpoint(filepath=os.path.join(location_save, 
													     'temp_mnist_weights.h5'),
                                   monitor='val_loss',
                                   verbose=verbose,
                                   save_best_only=True,
                                   save_weights_only=True,
                                   save_freq=save_freq
                                   )

# Fit model 
model.fit(x_train, 
		  y_train,
		  validation_data=(x_test, y_test),
		  batch_size=batch_size, 
		  epochs=2, 
		  callbacks=[model_checkpoint],
		  verbose=verbose)
```

**Comment by ntakouris** [comment]
The same incompatibility exists with other callbacks like reduce lr on plateau

**Comment by gattia** [comment]
@ntakouris thanks for adding that. 

@Saduf2019 can this be reopened? Or should I open a new issue?

**Comment by mcourteaux** [comment]
Tensorflow issue handling is a freaking joke. Nobody takes anything serious. All issues get closed by some stupid bot just because nobody takes time to look at them.
